### PR TITLE
fix(postgresql): Fix NullPointerException comparing or writing a null…

### LINKED ISSUE #677: AbstractDataType.compare and GenericEnumType.typeCast
> **Migrated from SourceForge** — [Bugs #433](https://sourceforge.net/p/dbunit/bugs/433/)
> Originally reported by **arrom** on 2019-07-11 18:42:40.269000
> Priority: `5`
> Assigned to **jeffjensen**
> SF status: pending

AbstractDataType.compare at line 82 calls GenericEnumType.typeCast
However,  GenericEnumType.typeCast does not check if object is null when it calls "return arg0.toString();", trhwoing a NullPointerException under certain conditions.

---

**jeffjensen** — 2019-07-12 00:52:12.499000
Thank you for reporting.

Are you able to make a merge request with the fix and tests?
If so, this page has helpful info to so so:
http://dbunit.sourceforge.net/devguide.html

### LINKED ISSUE #930: PostgreSQL uuid/inet/citext/enum columns NPE when comparing a null value against a non-null value
## Background

Found while implementing #574 (PostgresqlDataTypeFactory `json`/`jsonb` support). That issue's own history has a 2015 maintainer reply flagging the original patch attachment's `setSqlValue()` NPEing on a null value; tracing exactly why that class of bug is possible surfaced that four *already-shipped* `org.dbunit.ext.postgresql` classes have the same gap today, live on `main`.

## Root cause

`org.dbunit.dataset.datatype.AbstractDataType#compare(Object, Object)` - used by `Assertion`/`DbUnitAssert` whenever comparing an expected value against an actual value - calls `typeCast(o1)`/`typeCast(o2)` directly whenever the two values aren't `.equals()`, including the case where exactly one of them is null:

```java
if (areObjectsEqual(o1, o2)) { return 0; }
final Object value1 = typeCast(o1);
final Object value2 = typeCast(o2);
```

Four PostgreSQL `Types.OTHER` data types don't null-check before dereferencing in `typeCast(Object arg0)`, returning `arg0.toString()` unconditionally:

* `UuidType`
* `InetType`
* `CitextType`
* `GenericEnumType`

Additionally, none of the four null-check inside the private PGobject-building helper (`getUUID()`/`getInet()`/`getCitext()`/`getEnum()`) that `setSqlValue()` calls, so a direct `setSqlValue(null, ...)` call would also NPE. This is normally masked on the write path because `SimplePreparedStatement`/`PreparedBatchStatement#addValue()` intercept a null value and call `PreparedStatement#setNull(...)` directly before ever reaching `DataType#setSqlValue()` - but `AbstractDataType#compare()` has no equivalent guard, so the assertion path is genuinely reachable.

## Reproduction

Comparing an expected dataset row containing `null` for a `uuid`/`inet`/`citext`/enum column against an actual row containing a non-null value (or vice versa) via `Assertion.assertEquals`/`DbUnitAssert` throws `NullPointerException` instead of correctly reporting a value mismatch.

## Suggested fix

Null-check in `typeCast()` (returning `null` for a `null` input) and in `setSqlValue()` (bind SQL `NULL` before ever entering the PGobject-building path) for all four classes. `org.dbunit.ext.postgresql.JsonType`, added in #574 (branch `574-postgresql-json-type`, not yet merged), already implements both guards correctly and can be used as the reference pattern.

## Affected classes

* `org.dbunit.ext.postgresql.UuidType`
* `org.dbunit.ext.postgresql.InetType`
* `org.dbunit.ext.postgresql.CitextType`
* `org.dbunit.ext.postgresql.GenericEnumType`

### PR DESCRIPTION
**Pull request by jeffjensen**
… enum, uuid, inet, or citext value

GenericEnumType.typeCast() had no null guard, so AbstractDataType.compare() threw NullPointerException instead of reporting a mismatch whenever exactly one side of a compared pair was null. UuidType, InetType, and CitextType shared the same two-part gap: an unguarded typeCast(), plus a setSqlValue() override that bypasses typeCast() entirely to call a private PGobject- building helper that also dereferences the value unconditionally.

* Return null from typeCast() for a null input in all four classes, matching every other DataType implementation.
* Bind sql NULL from setSqlValue() before ever reaching the PGobject- building helper, matching the pattern JsonType already established.
* Add unit coverage per class: typeCast(null) returns null, compare(null, nonNullValue) no longer throws (the literal reported crash), and a Mockito-based setSqlValue(null, ...) proof that sql NULL is bound instead of thrown.
* Add PostgresqlNullableOtherTypesIT, round-tripping a null uuid/inet/citext row through CLEAN_INSERT against a live PostgreSQL 16 container. GenericEnumType is proven at the unit level only: a separate, pre-existing defect (issue 933, filed but not fixed here) leaves it unreachable for a real table column regardless of null handling, discovered while writing this test.
* Document the issue 933 caveat in postgresql.adoc's Known Quirks.

Refs: 677
Refs: 930

Claude-Session: https://claude.ai/code/session_01XBjSJ2bgVtwqyxGv2nvwP9

## Summary by Sourcery

Handle PostgreSQL enum, UUID, inet, and citext null values safely to avoid NullPointerExceptions and verify behavior with unit and integration tests.

Bug Fixes:
- Return null from GenericEnumType, UuidType, InetType, and CitextType typeCast() when given null and bind SQL NULL from setSqlValue() instead of throwing NullPointerException when handling nulls.

Documentation:
- Document the null-handling fix and GenericEnumType limitation in the PostgreSQL changelog and database documentation.

Tests:
- Add unit tests for GenericEnumType, UuidType, InetType, and CitextType to cover null typeCast(), compare() with null vs non-null values, and setSqlValue() binding SQL NULL without throwing.
- Add PostgresqlNullableOtherTypesIT to verify CLEAN_INSERT round-trips null uuid/inet/citext columns against a live PostgreSQL database without errors.

## Summary by CodeRabbit

- **Bug Fixes**
  - PostgreSQL UUID, network address, case-insensitive text, and enum values now handle `NULL` safely.
  - `NULL` values are correctly written to PostgreSQL without errors.
  - Comparisons and type conversions no longer throw exceptions for null values.

- **Documentation**
  - Expanded guidance on PostgreSQL enum handling and documented a known metadata-related limitation.

- **Tests**
  - Added unit and integration coverage for nullable PostgreSQL values and round-trip behavior.

### CODE REVIEW

File: src/main/java/org/dbunit/ext/postgresql/InetType.java [hunk 1]
```diff
@@ -59,13 +59,27 @@ public Object getSqlValue(int column, ResultSet resultSet) throws SQLException,
         return resultSet.getString(column);
     }
 
+    /**
+     * {@inheritDoc} Binds sql {@code NULL} when {@code uuid} is
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/main/java/org/dbunit/ext/postgresql/InetType.java]: Fixed in edf48362 — updated the Javadoc to say "the inet `value`" instead of `uuid`. Left the parameter name itself as `uuid` since that's pre-existing (shared by `UuidType`/`InetType`/`CitextType` since the original code), not something introduced by this PR.

File: src/main/java/org/dbunit/ext/postgresql/CitextType.java [hunk 2]
```diff
@@ -61,13 +61,27 @@ public Object getSqlValue(int column, ResultSet resultSet) throws SQLException,
         return resultSet.getString(column);
     }
 
+    /**
+     * {@inheritDoc} Binds sql {@code NULL} when {@code uuid} is
+     * {@code null}, instead of dereferencing it while building the
+     * PGobject.
+     */
     public void setSqlValue(Object uuid, int column,
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/main/java/org/dbunit/ext/postgresql/CitextType.java:69]: Fixed in edf48362 — updated the Javadoc on both CitextType and InetType to describe "the citext/inet `value`" instead of `uuid`. Left the parameter name itself as `uuid` since that's pre-existing (shared by `UuidType`/`InetType`/`CitextType` since the original code), not something introduced by this PR.
